### ISSUE TITLE
about the calculation of uncertainty

### ISSUE THREAD

**Issue by ShellingFord221** [kumar-shridhar/PyTorch-BayesianCNN#32]
Hi, in /Image Recognition/main_Bayes.py, line 226, I think there is something wrong about your calculation of the uncertainty.

In your code, test is also being done in the way of batches just like training. Assume that there are 5 instances in a batch, and num_samples is 10, then in line 224, results[0] should be a 50 * 1 tensor which contains max probabilities of 5 * 10 samples, and results[1] should also be a 50 * 1 tensor which contains predicted labels of these 50 samples. Then in line 226, you only append results[0][0] on conf, which is just the max probability of the first sample of the first instance in this batch. At last, after iterating all the batches, you calculate the uncertainty using conf, which only contains max probabilities of the first sample of the first instance in each batch. 

As far as I know, epistemic and aleatoric uncertainties are all based on the sampled results, which means that they are calculated on a probability matrix of 10 * 3 (assumed that there are 3 classes in the task) of each instance's sampled results. In other words, your should get all the probabilities after softmax in each sampling of one instance, then 10 sampled results form your p_hat matrix of just one instance. If you want to calculate the average uncertainty of all test instances, I think you should first calculate the sum of them, then calculate the average.

Am I correct? Thanks!

**Comment by kumar-shridhar** [comment]
Hi @ShellingFord221 ,

![image](https://user-images.githubusercontent.com/20888097/66357135-bc1efb00-e93b-11e9-9045-8175e4e9ba98.png)

**Comment by kumar-shridhar** [comment]
There is an updated repo here : [link](https://github.com/kumar-shridhar/PyTorch-Softplus-Normalization-Uncertainty-Estimation-Bayesian-CNN/blob/master/main.ipynb)

**Comment by ShellingFord221** [comment]
emmm... Why do you calculate the epistemic or aleatoric uncertainty by summing all the elements in the epistemic or aleatoric matrix?

**Comment by ShellingFord221** [comment]
Besides, in your new repo, num_samples = 1, which means you don't do stochastic forward pass when testing. I think in this case, you can't calculate epistemic or aleatoric uncertainty because they are all based on the sampled results. p_hat should be a numpy array of size (num_samples, num_classes) of only ONE test instance. Please see https://github.com/ykwon0407/UQ_BNN/issues/3#issuecomment-455121151

**Comment by kumar-shridhar** [comment]
Num of samples is a hyperparameter and can be changed. One is just an example and to make sure it fits in everyone's memory who is running it. 

I am summing it because it is a normalized way to calculate the uncertainty as described in the paper. We can also do it with the outputs from softmax which is described in the above function in the repo.
